replacy/suggestion.py

```diff
@@ -0,0 +1,306 @@
+                # to do: match again to get multi-token
+                try:
+                    # map ref to positive
+                    ref = len(pattern_ref) + ref
+                    refd_tokens = pattern_ref[ref]
+                    if len(refd_tokens):
+                        min_i = start + min(refd_tokens)
+                        max_i = start + max(refd_tokens)
+                        refd_text = doc[min_i : max_i + 1].text
+                    else:
+                        refd_text = None
+                except:
+                    warnings.warn(
+                        f"Ref matcher failed for span {doc[start:end]} and {pattern_ref}."
+                    )
+                    refd_text = doc[end + ref].text
+
+            if refd_text:
+                item_options = [refd_text]
+            else:
+                item_options = []
+
+        return item_options
+
+    def get_item_max_count(self, item, item_options):
+
+        # max count can be hard set in match_dict
+        max_count = item.get("MAX_COUNT", None)
+        if max_count:
+            return max_count
+
+        # can be soft set by default 
+        # but no more than possible - ex. list len
+        # or maximal ie. list len
+        if self.default_max_count:
+            max_count = min(self.default_max_count, len(item_options))
+        else:
+            max_count = len(item_options)
+
+        # if we don't want to guess max count
+        # to eliminate grammatical variants
+        # end here
+        if not self.filter_suggestions:
+            return max_count
+        
+        # if max count is not hard set
+        # try to lower max count in special cases (A - G)
+        # to eliminate non grammatical suggestions
+
+        # A. empty
+        # ex. []
+        if not len(item_options):
+            return 1
+
+        # B. contains non letters
+        # ex. ["", ","]
+        if not all([o.isalpha() for o in item_options]):
+            return 1
+
+        # C. is multi token
+        # ex. ["in a", "for"]
+        if max([len(o.split()) for o in item_options]) > 1:
+            return 1
+
+        # D. if inflection is set to tag - good
+        # other options - will always return many
+        if "INFLECTION" in item:
+            inflection = item.get("INFLECTION")
+            inflection_type = self.inflector.get_inflection_type(inflection)
+            if inflection_type != "tag":
+                return 1
+
+        # contains many options
+        # ex. ["eat", "walk"]
+        if len(item_options) > 1:
+
+            # E. contains words of the same lemma
+            # ex. [slow, slowly]
+            lemmas = set([])
+            for option in item_options:
+                option_lemmas = set(self.inflector.get_lemmas(option))
+                if len(lemmas & option_lemmas):
+                    return 1
+                lemmas = lemmas.union(option_lemmas)
```

Comment:
not a recommendation, but fun fact: this can be written `lemmas |= option_lemmas`

Reply:
:chefkiss: